### PR TITLE
move `modifiableLvalue` to `expressionsem.d`

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

**Comment by PetarKirov** [comment]
@thewilsonator update?

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -986,6 +986,181 @@ Expression resolvePropertiesOnly(Scope* sc, Expression e1)
     return e1;
 }
 
+Expression modifiableLvalue(Expression e, Scope* sc, Expression ebackup)
```
Comment (ibuclaw) [inline compiler/src/dmd/expressionsem.d:989]: `eorig` is probably a more accurate name.

**Comment by thewilsonator** [comment]
> @thewilsonator update?

got stuck and can't debug locally (because DMD doesn't build on M1). I think it might be something to do with casting and the visitor playing tricks. A better approach might be the classless explicit visitor that Walter has been using. There is an example of that working somewhere in the semantic that was a transformation similar to this (i.e. taking what was a virtual function in the class hierarchy and extracting it out). I don't recall where offhand.

This should still be done at some point.  Preferably by someone else :)

**Comment by PetarKirov** [comment]
> because DMD doesn't build on M1

That's odd - it does work for me on a M1 Pro macOS 14.0 machine for me using [Nix](https://nixos.org/). See for yourself:
https://asciinema.org/a/6QOXBKl94rh20zEJ8uC5VJfUG

I did have to use ldc and clang compiled for x86_64-darwin instead of aarch64-darwin though, but this works fine with Rosetta.

**Comment by PetarKirov** [comment]
> because DMD doesn't build on M1

@thewilsonator That's odd - it does work for me on a M1 Pro macOS 14.0 machine for me using [Nix](https://nixos.org/). See for yourself:
https://asciinema.org/a/6QOXBKl94rh20zEJ8uC5VJfUG

I did have to use ldc and clang compiled for x86_64-darwin instead of aarch64-darwin though, but this works fine with Rosetta.

**Comment by thewilsonator** [comment]
Superseded by https://github.com/dlang/dmd/pull/15781

**Pull request closed without merging**